### ISSUE TITLE
InunMap executable not created during compile

### ISSUE THREAD

**Issue by amoodie** [dtarb/TauDEM#234]
## Background

Trying to get the InunMap function working on linux (Google Colab specifically). As a result, I'm compiling from source, following the instructions in the readme. This works for the most part, and executables wind up in `usr/local/taudem` as expected.

## Problem

However, the `inunmap` executable is not created. This, I guess, may be expected given the source files [are not listed in any targets in the `CMakeLists.txt` file](https://github.com/dtarb/TauDEM/blob/98137bb6541a0d0077a9c95becfed4e56d0aa0ac/src/CMakeLists.txt).

## Attempt

I have made some [changes to the file, on a branch here](https://github.com/amoodie/TauDEM/pull/1/files), that I thought might create the target to compile the `inunmap` executable, which seemed to work. By seemed, I mean a target for `inunmap` is created in the `Makefile` after `cmake ..`, and a folder `inunmap.dir` is created with some (??) source files (image below).

![image](https://user-images.githubusercontent.com/8801322/141041383-c0f23451-086a-4c03-99dd-760294b2285d.png)

The problem is that when I actually try to compile and install (i.e., `make && make install`), I get an error about the `nc_open` symbol not found: 
```
-- Configuring done
-- Generating done
-- Build files have been written to: /content/TauDEM/src/build
[  1%] Linking CXX executable inunmap
CMakeFiles/inunmap.dir/build.make:131: recipe for target 'inunmap' failed
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/inunmap.dir/all' failed
Makefile:129: recipe for target 'all' failed
mkdir: cannot create directory ‘build’: File exists
/usr/bin/ld: CMakeFiles/inunmap.dir/InunMap.cpp.o: undefined reference to symbol 'nc_open'
//usr/lib/x86_64-linux-gnu/libnetcdf.so.13: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [inunmap] Error 1
make[1]: *** [CMakeFiles/inunmap.dir/all] Error 2
make: *** [all] Error 2
```

I _guess_ this is from a netcdf C library not being found. Note, there is no netcdf library in the above image, but should there be? I'm really pushing my knowledge of C and makefiles with this, so please forgive me if my terminology is wrong. Hopefully this is clear enough to get some help though.

Do you know why the netcdf library is not found? Is this library distributed with TauDEM or was I supposed to install it myself? It's entirely possible I have not installed it at as a shared library at all -- do you recommend where to get it from? Any pointers or ideas would be much appreciated. Thank you 🙏

**Comment by dtarb** [comment]
You are on the right track.  Inunmap is the only function that uses netcdf.  It is not on the make targets because we have not had time to sort this problem out.  

Netcdf is available from Unidata.  See https://www.unidata.ucar.edu/software/netcdf/.  Hopefully some information there can help you out.

I also noticed that in the Windows compilation the netcdf.h file is in a GDAL folder, which suggests that GDAL may provide the functions needed.  Setting up paths to find this thus also has a possibility of working.

I hope that either of these suggestions can move you along.  I would actually prefer to remove the netcdf dependency and use text files instead, but have not had time to do this.  It requires C code changing.

Also, given that you are working in Google Colab, you might try swing for the fence and !pip install netcdf4 at the beginning and see if you get an installation that you can link TauDEM to. 

Good luck!

**Comment by amoodie** [comment]
Thanks, I've installed the netcdf library with `sudo apt install netcdf-dev` which places the netCDF C header file (`netcdf.h`) in `/usr/include` (I verified, it is there). Compilation still fails with the same error (`undefined reference to symbol 'nc_open'`)

I checked the makefile [C compiler options and checked the output of `nc-config --cflags`](https://github.com/dtarb/TauDEM/blob/98137bb6541a0d0077a9c95becfed4e56d0aa0ac/src/makefile#L96) is indeed as expected: 
```
$ nc-config --cflags
-I/usr/include -I/usr/include/hdf5/serial
```

IS there some cache that needs to be cleared with the makefile or something? Sorry, I'm again pushing my understanding of C here... Any suggestions appreciated, but I acknowledge we're in some "unsupported feature" territory now.


Note: `pip` installing did not work (C header still not found). I have not tried linking GDAL libraries, and to be honest, I wouldn't really know where to start on that.

**Comment by amoodie** [comment]
Alright I've worked out a way to make this work.

Installing netcdf libraries with 
```
sudo apt install libhdf5-serial-dev netcdf-bin libnetcdf-dev libnetcdff-dev --yes
```
Then using the [changes on the branch here](https://github.com/amoodie/TauDEM/pull/1/files) allowed me to compile `inunmap` successfully. Most significant changes were to the `CMakeLists.txt` file, and a `FindXXXX.cmake` file for the netcdf library ([stolen from Jed Brown's repo here](https://github.com/jedbrown/cmake-modules/blob/master/FindNetCDF.cmake)). I guess these paths can probably be specified directly too somehow, but I didn't know how.

Sharing this in case it's helpful for any of your future development, and it will at least allow Paola to move forward with her lab exercise. Looks like Yan also sent Paola a cleaner way to compile and install on Colab, which will be helpful in the future.

For what it's worth to one of the above suggestions, I did look in `/usr/include/gdal` (was installed with `sudo apt install gdal-bin libgdal-dev`) and there is *not* a copy of `netcdf.h` in that folder.
